### PR TITLE
Faster merging of hyper c arrays.

### PR DESCRIPTION
**Pull request by cmdevries**
I tried implementing 3 different approaches to merging the arrays. This was by
far the fastest at (0.01ms to 0.02ms for the union benchmark). It seems
when the ternary operator is inside the inner loop compiler does something that
is fast such as an operation with no branching.

The other approaches were,

* processing a 64-bit integer at a time and unpacking the bytes (0.16ms)
* branching rather than a ternary operator for the max value which is
  similar to the original code (0.10ms to 0.40ms)

Benchmark before,

module       P        card   fill      bytes  insert us   union ms    card ms    json ms
hyper_carray 15          1   0.00      32784       1.80       0.10       0.33       0.21
hyper_carray 15        100   0.00      32784       0.54       0.11       0.34       0.28
hyper_carray 15        500   0.02      32784       0.50       0.10       0.38       0.47
hyper_carray 15       1000   0.03      32784       0.51       0.11       0.38       0.74
hyper_carray 15       2500   0.07      32784       0.50       0.11       0.43       1.41
hyper_carray 15       5000   0.14      32784       0.51       0.12       0.51       2.38
hyper_carray 15      10000   0.26      32784       0.51       0.14       0.65       4.17
hyper_carray 15      15000   0.37      32784       0.51       0.16       0.82       5.32
hyper_carray 15      25000   0.53      32784       0.51       0.20       1.05       6.10
hyper_carray 15      50000   0.78      32784       0.51       0.26       1.31       5.77
hyper_carray 15     100000   0.95      32784       0.51       0.33       1.65       5.18
hyper_carray 15    1000000   1.00      32784       0.53       0.43       1.83       5.23

Benchmark after,

module       P        card   fill      bytes  insert us   union ms    card ms    json ms
hyper_carray 15          1   0.00      32784       2.30       0.02       0.32       0.22
hyper_carray 15        100   0.00      32784       0.53       0.02       0.33       0.22
hyper_carray 15        500   0.02      32784       0.51       0.02       0.35       0.43
hyper_carray 15       1000   0.03      32784       0.50       0.02       0.37       0.68
hyper_carray 15       2500   0.07      32784       0.51       0.02       0.42       1.34
hyper_carray 15       5000   0.14      32784       0.50       0.01       0.50       2.33
hyper_carray 15      10000   0.26      32784       0.52       0.01       0.65       4.10
hyper_carray 15      15000   0.37      32784       0.51       0.02       0.79       5.33
hyper_carray 15      25000   0.53      32784       0.51       0.02       1.01       6.11
hyper_carray 15      50000   0.78      32784       0.51       0.02       1.30       5.77
hyper_carray 15     100000   0.95      32784       0.51       0.02       1.63       5.23
hyper_carray 15    1000000   1.00      32784       0.53       0.02       1.82       5.32

### CODE REVIEW

File: c_src/hyper_carray.c [hunk 1]
```diff
@@ -78,13 +81,12 @@ static void carray_alloc(unsigned int precision,
...
-static inline void carray_merge_item(struct hyper_carray *restrict arr,
+static inline void carray_merge_item(carray_ptr arr,
 				     unsigned int index,
 				     unsigned int value)
 {
-	char *item = arr->items + index;
-	if (value > *item)
-		*item = value;
+	uint8_t *item = arr->items + index;
+    *item = (value > *item) ? value : *item;
```
Comment (cmdevries) [inline c_src/hyper_carray.c]: need to fix indenting to be consistent

File: c_src/hyper_carray.c [hunk 2]
```diff
@@ -151,39 +153,52 @@ static ERL_NIF_TERM max_merge(ErlNifEnv * env, int argc,
...
+	carray_ptr first = NULL;
 	HYPER_CARRAY_OR_BADARG(head, first);
 
+	carray_ptr acc = NULL;
 	carray_alloc(first->precision, &acc);
 	memcpy(acc->items, first->items, acc->size);
 
-	unsigned int nitems = 0;
+	carray_ptr * arrays = (carray_ptr *)enif_alloc(narrays * sizeof(carray_ptr));
+    arrays[0] = first;
```
Comment (cmdevries) [inline c_src/hyper_carray.c]: indenting

File: c_src/hyper_carray.c [hunk 3]
```diff
@@ -151,39 +153,52 @@ static ERL_NIF_TERM max_merge(ErlNifEnv * env, int argc,
...
-
 		continue;
 
 	      dealloc_and_badarg:
+        enif_free((void *)arrays);
 		dtor(env, acc);
 		return enif_make_badarg(env);
 	}
 
+    // Merge arrays
```
Comment (cmdevries) [inline c_src/hyper_carray.c]: indenting again

File: c_src/hyper_carray.c [hunk 4]
```diff
@@ -40,9 +41,11 @@ struct hyper_carray {
 	/*
 	 * Array of items each one byte in size.
 	 */
-	char *items;
+	uint8_t *items;
 };
 
+typedef struct hyper_carray *restrict carray_ptr;
```
Comment (johannesh) [inline c_src/hyper_carray.c]: :+1:

File: c_src/hyper_carray.c [hunk 5]
```diff
@@ -151,39 +152,54 @@ static ERL_NIF_TERM max_merge(ErlNifEnv * env, int argc,
...
 	memcpy(acc->items, first->items, acc->size);
 
-	unsigned int nitems = 0;
+	carray_ptr *arrays =
+	    (carray_ptr *) enif_alloc(narrays * sizeof(carray_ptr));
+	arrays[0] = first;
+
+	// Validate arrays
 	for (int i = 1; i < narrays; ++i) {
+		void *tmp = NULL;
```
Comment (johannesh) [inline c_src/hyper_carray.c]: :+1:

File: c_src/hyper_carray.c [hunk 6]
```diff
@@ -151,39 +152,54 @@ static ERL_NIF_TERM max_merge(ErlNifEnv * env, int argc,
...
+	carray_ptr first = NULL;
 	HYPER_CARRAY_OR_BADARG(head, first);
 
+	carray_ptr acc = NULL;
 	carray_alloc(first->precision, &acc);
 	memcpy(acc->items, first->items, acc->size);
 
-	unsigned int nitems = 0;
+	carray_ptr *arrays =
+	    (carray_ptr *) enif_alloc(narrays * sizeof(carray_ptr));
```
Comment (johannesh) [inline c_src/hyper_carray.c]: Allocated for?

**Pull request merged**